cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java

```diff
@@ -0,0 +1,178 @@
+  );
+
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+  private final int batchSize;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+  private volatile boolean hasV1Instance;
+
+  MetadataMigrator(CConfiguration cConf, DatasetFramework dsFramework, TransactionSystemClient txClient) {
+    this.dsFramework = dsFramework;
+    this.batchSize = cConf.getInt(Constants.Metadata.MIGRATOR_BATCH_SIZE);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(new SystemDatasetInstantiator(dsFramework),
+                                                                   txClient, NamespaceId.SYSTEM,
+                                                                   Collections.emptyMap(), null, null)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting Metadata Migrator Service.");
+  }
+
+  @Override
+  public void run() {
+    runThread = Thread.currentThread();
+
+    try {
+      for (int i = 0; i < DATASET_IDS.size(); i = i + 2) {
+        // assume we already have v1 metadata table instance
+        hasV1Instance = true;
+
+        while (!stopped && hasV1Instance) {
+          try {
+            final int index = i;
+            // This thread migrates metadata in batches. It does following operations in a single transaction:
+            // 1.) Scan V1 metadata table for value and history rows in batch of batchSize.
+            // 2.) Write scanned MetadataEntries to V2 table.
+            // 3.) Delete successfully written metadata entries from V1 table.
+            Transactionals.execute(transactional, context -> {
+              if (!dsFramework.hasInstance(DATASET_IDS.get(index))) {
+                hasV1Instance = false;
+                return;
+              }
+
+              MetadataDataset v1 = getMetadataDataset(context, DATASET_IDS.get(index));
+              MetadataDataset v2 = getMetadataDataset(context, DATASET_IDS.get(index + 1));
+
+              // All the metadata entries are written using setProperty because it does not modify the MetadataEntry
+              // keys
+              List<KeyValue<Long, MetadataEntry>> entries = v1.scanOrDeleteFromV1Table(batchSize, false);
+
+              if (entries.isEmpty()) {
+                // All the value and history rows have been migrated so stop this thread and drop V1 MetadataDataset
+                dropV1MetadataDataset(index);
+                LOG.debug("Migration for dataset {} is complete. This dataset is dropped.", DATASET_IDS.get(index));
+                return;
+              }
+
+              v2.writeUpgradedRow(entries);
+              // We do not need to keep checkpoints. Instead, we will just delete scanned rows from metadata dataset
+              v1.scanOrDeleteFromV1Table(entries.size(), true);
```

Comment:
when deleting, do you actually have to scan again? Or can you just do a batch delete of all row keys that you scanned in the first round? That might require remembering the actual keys for each entry, but seems more efficient to me.